org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/export/CsarExporter.java

```diff
@@ -214,6 +247,77 @@ public String writeCsar(IRepository repository, DefinitionsChildId entryId, Outp
+        TServiceTemplate serviceTemplate = repository.getElement(entryId);
+
+        @NonNull List<TNodeTemplate> nodeTemplates = serviceTemplate.getTopologyTemplate().getNodeTemplates();
+
+        TTopologyTemplate topologyTemplateCopy = serviceTemplate.getTopologyTemplate();
+
+        nodeTemplates.forEach(tNodeTemplate -> {
+            boolean ubuntuFlag = false;
+            boolean dockerFlag = false;
+            if (tNodeTemplate.getName().startsWith("Ubuntu-VM")) {
+                ubuntuFlag = true;
+            }
+
+            if (tNodeTemplate.getDeploymentArtifacts() != null) {
+                @NonNull List<TDeploymentArtifact> deploymentArtifacts = tNodeTemplate.getDeploymentArtifacts().getDeploymentArtifact();
+
+                for (TDeploymentArtifact deploymentArtifact :
+                    deploymentArtifacts) {
+
+                    if (deploymentArtifact.getArtifactType().getLocalPart().equals("DockerContainerArtifact")) {
+                        dockerFlag = true;
+                    }
+                }
+            }
+
+            if (dockerFlag || ubuntuFlag) {
+                TNodeTemplate duplicatedNode = BackendUtils.clone(tNodeTemplate);
+                duplicatedNode.setId(tNodeTemplate.getName());
+                duplicatedNode.setName(tNodeTemplate.getName());
+
+                QName tNodeQtype = tNodeTemplate.getTypeAsQName();
+                QName qName = new QName(tNodeQtype.getNamespaceURI(), tNodeQtype.getLocalPart() + "-self", tNodeQtype.getPrefix());
+
+                duplicatedNode.setType(qName);
+
+                if (dockerFlag) {
+                    @Nullable QName artifactRef = duplicatedNode.getDeploymentArtifacts().getDeploymentArtifact().get(0).getArtifactRef();
+                    String artifactTemplateName = "{http://opentosca.org/artifacttemplates}" + artifactRef.getLocalPart() + "-self";
+
+                    ArtifactTemplateId newArtifactTemplateId = new ArtifactTemplateId(QName.valueOf(artifactTemplateName));
+
+                    TArtifactTemplate artifactTemplate = repository.getElement(newArtifactTemplateId);
+
+                    List<TDeploymentArtifact> artifactTemplateList = new ArrayList<>();
+                    artifactTemplateList.add(new TDeploymentArtifact.Builder(artifactTemplate.getName(), artifactTemplate.getType())
+                        .setArtifactRef(QName.valueOf("{http://opentosca.org/artifacttemplates}" + artifactTemplate.getId())).build());
+
+                    TDeploymentArtifacts DAs = new TDeploymentArtifacts.Builder(artifactTemplateList).build();
+
+                    duplicatedNode.setDeploymentArtifacts(DAs);
+                }
+
+                topologyTemplateCopy.getNodeTemplateOrRelationshipTemplate().add(duplicatedNode);
+                topologyTemplateCopy.getNodeTemplateOrRelationshipTemplate().remove(tNodeTemplate);
```

Comment:
I would change the order here, first remove it, then add it again because it may happen that it deletes the element based on the `equals` method and I think that they are equal if the IDs are equal, but not quite sure. Just to be save.